### ISSUE TITLE
Errors during RESET step

### ISSUE THREAD

**Issue by saty9** [JuulLabs-OSS/mcumgr-android#87]
I have been receiving 2 errors during the reset state:

one is: `io.runtime.mcumgr.exception.McuMgrException: UNKNOWN(-5)`
the other is: `io.runtime.mcumgr.exception.McuMgrException: io.runtime.mcumgr.ble.callback.TransactionTimeoutException: Transaction 2 timed out without receiving a response`

I think both are due to the device disconnecting at reset and then taking too long to come back up.
If a subsequent upload of the same firmware is attempted it goes straight to confirm after validate so I think the library is just not waiting long enough after a reset/not retrying if the first connection fails.

**Comment by saty9** [comment]
fixed by increasing the expected swap time

**Comment by bgiori** [comment]
Hi @saty9,

Thanks for submitting this issue. 

```
io.runtime.mcumgr.exception.McuMgrException: io.runtime.mcumgr.ble.callback.TransactionTimeoutException: Transaction 2 timed out without receiving a response
```

There are some bugs in the `0.12.0-beta` versions (assuming you are using one of those for faster DFU) with a new implementation of the BLE transport layer. As of now I am still working them out, so you may want to switch to 0.11.0 until the non-beta release gets tagged. If 0.11.0 doesn't work for you please let me know .

```
io.runtime.mcumgr.exception.McuMgrException: UNKNOWN(-5)
```

This seems a bit problematic, as there is no code -5. I would suppose that some non-standard data is getting received and parsed out. Does this happen often?

> fixed by increasing the expected swap time but would be helpful if it gave a clearer/ more consistent error message

For sure, this is a good callout. I can wrap any exceptions received during the reset step in an exception with additional info about the amount of time waited and a message suggesting increasing the swap time. Should be less confusing for developers.

**Comment by philips77** [comment]
-5 error may be coming from [here](https://github.com/NordicSemiconductor/Android-BLE-Library/blob/1b836ad846ef86fed35fc93807ff8e1056afd57f/ble/src/main/java/no/nordicsemi/android/ble/callback/FailCallback.java#L34).

**Comment by saty9** [comment]
@bgiori I encountered these errors in 0.11 and tried 0.12 in the hope it would make them go away but they also appeared then too

**Comment by bgiori** [comment]
> -5 error may be coming from [here](https://github.com/NordicSemiconductor/Android-BLE-Library/blob/1b836ad846ef86fed35fc93807ff8e1056afd57f/ble/src/main/java/no/nordicsemi/android/ble/callback/FailCallback.java#L34).

That would make sense. Would be worth cleaning up the error message, via wrapping he exception in a failure rather than using the status as an mcumgr error code. 

> @bgiori I encountered these errors in 0.11 and tried 0.12 in the hope it would make them go away but they also appeared then too

The `TransactionTimeoutException` is specific to 0.12.0-beta releases and was caused by a bug in one of the early releases. The bug has since been fixed in the most recent 0.12.0-beta4 release.

---------------

My conclusion is that we need better and more consistent exceptions for some error cases with regards to resets during DFU.
